java/statestore-committer/src/main/java/sleeper/statestore/committer/MultiThreadedStateStoreCommitter.java

```diff
@@ -0,0 +1,418 @@
+                    .map(message -> {
+                        LOGGER.trace("Received message: {}", message);
+                        StateStoreCommitRequest request = this.serDe.fromJson(message.body());
+                        LOGGER.trace("Received request: {}", request);
+                        return new StateStoreCommitRequestWithSqsReceipt(request, message.receiptHandle());
+                    })
+                    .collect(Collectors.groupingBy(request -> request.getCommitRequest().getTableId()));
+
+                // Try to make sure there is going to be enough heap space available to process these commits
+                this.ensureEnoughHeapSpaceAvailable(messagesByTableId.keySet());
+
+                messagesByTableId.entrySet().forEach(tableMessages -> {
+                    String tableId = tableMessages.getKey();
+                    List<StateStoreCommitRequestWithSqsReceipt> requestsWithHandle = tableMessages.getValue();
+                    LOGGER.info("Received {} requests for table: {}", requestsWithHandle.size(), tableId);
+
+                    // Wait until processing of previous commits for this table has finished
+                    if (this.tableFutures.containsKey(tableId)) {
+                        try {
+                            this.tableFutures.get(tableId).get();
+                        } catch (Exception e) {
+                            throw new IllegalStateException("Exception was thrown during the processing of the previous batch of commits for table " + tableId + ":", e);
+                        }
+                    }
+
+                    // Apply the commits for each table on a separate thread
+                    CompletableFuture<Instant> task = CompletableFuture.supplyAsync(() -> {
+                        this.processMessagesForTable(tableId, requestsWithHandle);
+                        return Instant.now();
+                    });
+
+                    this.tableFutures.put(tableId, task);
+                    this.processedTableOrder.remove(tableId);
+                    this.processedTableOrder.add(tableId);
+                });
+            }
+        } catch (Exception e) {
+            LOGGER.error("Caught exception, starting graceful shutdown:", e);
+            err = e;
+        } finally {
+            long pendingTaskCount = tableFutures.values().stream().filter(task -> !task.isDone()).count();
+            if (pendingTaskCount > 0) {
+                LOGGER.info("Requests for {} tables are still being processed, waiting for them to finish...", pendingTaskCount);
+            }
+            tableFutures.values().stream().map(task -> task.join()).collect(Collectors.toList());
+            LOGGER.info("All pending requests have been actioned");
+            if (err != null) {
+                throw err;
+            }
+        }
+    }
+
+    private void ensureEnoughHeapSpaceAvailable(Set<String> requiredTableIds) {
```

Comment:
This seems like it could be moved into StateStoreProvider? The processedTableOrder field seems to duplicate the field `StateStoreProvider.tableIds`.